### ISSUE TITLE
Google+ API Being Deprecated

### ISSUE THREAD

**Issue by ottboy4** [alvarosanchez/grails-spring-security-rest#395]
The oauth login through Google appears to rely on the Google+ API which is being deprecated by  March 7, 2019.
See: https://developers.google.com/+/web/signin/

Are there plans for migrating this project to Google's new oauth solution? Otherwise, my Google based login will stop working when the Google+ API is shut down.

**Comment by ottboy4** [comment]
This issue comes from pac4j since that is the library that implements the Google OAuth login.

It is currently being worked on and fixed there. 
https://github.com/pac4j/pac4j/issues/1228

**Comment by ottboy4** [comment]
@ivarprudnikov the plugin is using an outdated version of pac4j. So what you are saying makes sense, except the details are a little bit different since it is an old version.

Here is my implementation which is working currently for me. It uses an updated profile URL, but does not support emails (plural, email is still supplied) and birthday.

Here is my Google2Client override. I embedded the other helper classes since this code is only needed till the plugin is updated to the new (once fixed) version of pac4j.

```
import com.fasterxml.jackson.databind.JsonNode
import org.pac4j.core.profile.AttributesDefinition
import org.pac4j.core.profile.converter.Converters
import org.pac4j.oauth.client.Google2Client
import org.pac4j.oauth.profile.JsonHelper
import org.pac4j.oauth.profile.OAuthAttributesDefinition
import org.pac4j.oauth.profile.google2.Google2Email
import org.pac4j.oauth.profile.google2.Google2Profile
import org.scribe.model.Token

class OverrideGoogle2Client extends Google2Client {

    OverrideGoogle2Client() {
    }

    OverrideGoogle2Client(String key, String secret) {
        super(key, secret)
    }

    @Override
    protected String getProfileUrl(Token accessToken) {
        return "https://www.googleapis.com/oauth2/v3/userinfo"
    }

    @Override
    protected Google2Profile extractUserProfile(String body) {
        OverrideGoogle2Profile profile = new OverrideGoogle2Profile()
        JsonNode json = JsonHelper.getFirstNode(body)
        if (json != null) {
            profile.setId(JsonHelper.get(json, 'sub')) // id stored in "sub" field
            for (String attribute : OverrideGoogle2AttributesDefinition.INSTANCE.getPrincipalAttributes()) {
                profile.addAttribute(attribute, JsonHelper.get(json, attribute))
            }
        }
        return profile
    }

    static class OverrideGoogle2Profile extends Google2Profile {

        @Override
        protected AttributesDefinition getAttributesDefinition() {
            return OverrideGoogle2AttributesDefinition.INSTANCE
        }

        @Override
        String getEmail() {
            return getAttribute(OverrideGoogle2AttributesDefinition.EMAIL)
        }

        @Override
        String getFirstName() {
            return getAttribute(OverrideGoogle2AttributesDefinition.GIVEN_NAME)
        }

        @Override
        String getFamilyName() {
            return getAttribute(OverrideGoogle2AttributesDefinition.FAMILY_NAME)
        }

        @Override
        String getDisplayName() {
            return getAttribute(OverrideGoogle2AttributesDefinition.NAME)
        }

        @Override
        Locale getLocale() {
            return (Locale) getAttribute(OverrideGoogle2AttributesDefinition.LOCALE)
        }

        @Override
        String getPictureUrl() {
            return getAttribute(OverrideGoogle2AttributesDefinition.PICTURE)
        }

        @Override
        String getProfileUrl() {
            return getAttribute(OverrideGoogle2AttributesDefinition.PROFILE)
        }

        Date getBirthday() {
            return null // not supported
        }

        List<Google2Email> getEmails() {
            return null // not supported
        }
    }

    static class OverrideGoogle2AttributesDefinition extends OAuthAttributesDefinition {

        static final OverrideGoogle2AttributesDefinition INSTANCE = new OverrideGoogle2AttributesDefinition()

        static final String GENDER = 'gender'
        static final String NAME = 'name'
        static final String GIVEN_NAME = 'given_name'
        static final String FAMILY_NAME = 'family_name'
        static final String PROFILE = 'profile'
        static final String PICTURE = 'picture'
        static final String LOCALE = 'locale'
        static final String EMAIL = 'email'

        OverrideGoogle2AttributesDefinition() {
            addAttribute(GENDER, Converters.genderConverter)
            addAttribute(NAME, Converters.stringConverter)
            addAttribute(GIVEN_NAME, Converters.stringConverter)
            addAttribute(FAMILY_NAME, Converters.stringConverter)
            addAttribute(PROFILE, Converters.stringConverter)
            addAttribute(PICTURE, Converters.stringConverter)
            addAttribute(LOCALE, Converters.localeConverter)
            addAttribute(EMAIL, Converters.stringConverter)
        }
    }
}
```

**Comment by francoiskha** [comment]
Hi, [this has been fixed in pac4j 3.6+](https://github.com/alvarosanchez/grails-spring-security-rest/issues/395)

**Comment by ivarprudnikov** [comment]
Stating the obvious 
Original pac4j issue was closed on Feb 7, 2019 in `v3.6.0` as mentioned by @francoiskha :
https://github.com/pac4j/pac4j/issues/1228

Whilst this plugin still depends on pac4j v2.2.1 from Dec 13, 2017 :
https://github.com/alvarosanchez/grails-spring-security-rest/blob/develop/spring-security-rest/build.gradle#L2

**Comment by alvarosanchez** [comment]
Fixed via #416
